src/rust/engine/src/interning.rs

```diff
@@ -43,14 +43,32 @@ impl Interns {
     Interns::default()
   }
 
+  pub fn insert_product(&mut self, v: Value) -> Key {
+    let ProductTypeId { hash, type_id } = externs::product_type(&v);
+    let mut inserted = false;
+    let id_generator = self.id_generator;
+    let key = *self
+      .forward
+      .entry(InternKey(hash, v.clone()))
+      .or_insert_with(|| {
+        inserted = true;
+        Key::new(id_generator, type_id)
+      });
+    if inserted {
+      self.reverse.insert(key, v);
+      self.id_generator += 1;
+    }
+    key
+  }
+
   pub fn insert(&mut self, v: Value) -> Key {
-    let ident = externs::identify(&v);
-    let type_id = ident.type_id;
+    let type_id = externs::get_type_for(&v);
```

Comment:
It looks like `externs::get_type_for` does need to exist as distinct from `identify`. But I think that this path is hot enough that it would make sense to have `identify` continue to return the type id so that we can avoid calling two python methods during `fn insert`.

Reply:
Will do this!

Reply:
Done!